platform/cxxsupport/mstd_span

```diff
@@ -0,0 +1,365 @@
+
+    static constexpr index_type extent = Extent;
+
+    // Constructors, copy and assignment
+    template<size_t E = Extent,
+        typename mstd::enable_if_t<E == dynamic_extent, int> = 0>
+    constexpr span() noexcept
+    {}
+
+    template<class It, size_t E = Extent,
+        typename mstd::enable_if_t<mstd::is_convertible<
+        remove_reference_t<mstd::iter_reference_t<It>>(*)[],
+        ElementType(*)[]>::value, int> = 0>
+    constexpr span(It ptr, index_type count) :
+        _storage(ptr, count)
+    {
+        MBED_ASSERT(extent == dynamic_extent || extent == count);
+    }
+
+    template<class It, typename mstd::enable_if_t<mstd::is_convertible<
+        remove_reference_t<mstd::iter_reference_t<It>>(*)[],
+        ElementType(*)[]>::value, int> = 0>
+    constexpr span(It first, It last) :
+        _storage(first, last - first)
+    {
+        MBED_ASSERT(first <= last);
+        MBED_ASSERT(extent == dynamic_extent || extent == last - first);
+        MBED_ASSERT(extent == 0 || nullptr != first);
+    }
+
+    template<size_t N>
+    constexpr span(type_identity_t<element_type> (&arr)[N],
+            typename mstd::enable_if_t<(Extent == dynamic_extent || Extent == N) &&
+            mstd::is_convertible<remove_pointer_t<decltype(mstd::data(arr))>(*)[],
+            ElementType(*)[]>::value, int> = 0) noexcept:
+        _storage(arr, N)
+    {}
+
+    template<class T, size_t N>
+    constexpr span(array<T, N>& arr,
+            typename mstd::enable_if_t<(Extent == dynamic_extent || Extent == N) &&
+            mstd::is_convertible<remove_pointer_t<decltype(mstd::data(arr))>(*)[],
+            ElementType(*)[]>::value, int> = 0) noexcept:
+        _storage(arr.data(), N)
+    {}
+
+    template<class T, size_t N>
+    constexpr span(const array<T, N>& arr,
+            typename mstd::enable_if_t<(Extent == dynamic_extent || Extent == N) &&
+            mstd::is_convertible<remove_pointer_t<decltype(mstd::data(arr))>(*)[],
+            ElementType(*)[]>::value, int> = 0) noexcept:
+        _storage(arr.data(), N)
+    {}
+
+    template<class R,
+    typename mstd::enable_if_t<detail::is_container<R>::value &&
+            detail::is_compatible<R&, ElementType>::value, int> = 0>
+    constexpr span(R&& r) :
+        _storage( mstd::data( r ), mstd::size( r ))
+    {
+        MBED_ASSERT(extent == dynamic_extent || extent == mstd::size( r ));
+    }
+
+    constexpr span(const span& other) noexcept = default;
+
+    template<class OtherElementType, size_t OtherExtent,
+    typename mstd::enable_if_t<(Extent == dynamic_extent || OtherExtent == Extent) &&
+        mstd:is_convertible<OtherElementType(*)[], ElementType(*)[]>::value, int> = 0>
+    constexpr span(const span<OtherElementType, OtherExtent>& s) noexcept:
+        _storage(s.data(), s.size())
+    {}
+
+    ~span() noexcept = default;
+
+    constexpr span& operator=(const span& other) noexcept = default;
```

Comment:
I think there's an astyle exemption for the cxxsupport directory, because it can't handle all the template stuff, but still I believe it's fairly consistent on Mbed `span &other` style.

Reply:
I don't really know what you expect of me here ;)

Reply:
Mbed-style spacing of pointer/reference operators

    constexpr span &operator=(const span &other) noexcept = default;